src/bootloaders/shim-systemd.c

```diff
@@ -111,40 +114,46 @@ __cbm_export__ const BootLoader
         SYSTEMD_SRC_DIR                                                                            \
         "/"                                                                                        \
         "systemd-boot" EFI_SUFFIX
-#define DST_DIR "/EFI/" KERNEL_NAMESPACE
-#define SHIM_DST                                                                                   \
-        DST_DIR                                                                                    \
-        "/"                                                                                        \
-        "bootloader" EFI_SUFFIX
-#define SYSTEMD_DST                                                                                \
-        DST_DIR                                                                                    \
-        "/"                                                                                        \
-        "loader" EFI_SUFFIX
-#define KERNEL_DST_DIR DST_DIR "/kernel"
-#define SYSTEMD_CONFIG_DIR "/loader"
-#define SYSTEMD_CONFIG SYSTEMD_CONFIG_DIR "/loader.conf"
-#define SYSTEMD_ENTRIES SYSTEMD_CONFIG_DIR "/entries"
-
-/* path to fallback bootloader: /EFI/Boot/BOOT(X64|IA32).EFI */
-#define EFI_FALLBACK_DIR "/EFI/Boot"
-#define EFI_FALLBACK_PATH                                                                          \
-        EFI_FALLBACK_DIR                                                                           \
-        "/"                                                                                        \
-        "BOOT" EFI_SUFFIX
+
+/* these three path components needs to be probed. they are used to copy files
+ * onto ESP which uses FAT. on actual FAT, use of ALL CAPS is enough to
+ * construct usable EFI paths. however, probing is needed to comply with the
+ * tests. */
+#define ESP_EFI "EFI" /* /EFI on ESP */
+#define ESP_BOOT "BOOT" /* BOOT component in /EFI/Boot */
+#define EFI_FALLBACK "BOOT" EFI_SUFFIX_U /* e.g. BOOTX64.EFI */
+
+/* these path components can be used as-is, no need to probe */
+#define SHIM_DST "bootloader" EFI_SUFFIX
+#define SYSTEMD_DST "loader" EFI_SUFFIX
+#define KERNEL_DST_DIR "kernel"
+#define SYSTEMD_CONFIG_DIR "loader"
+#define SYSTEMD_ENTRIES_DIR "entries"
+
 
 static char *shim_src;
-static char *shim_dst_host; /* as accessible by the CMB for file ops. */
-static char *shim_dst_esp;  /* absolute location of shim on the ESP. */
 static char *systemd_src;
```

Comment:
All this globally defined stuff violates how the other bootloader systems are implemented, globals are hard to track